### PR TITLE
ENH: Add optimize.root_scalar() as a universal dispatcher for 1d root-finders

### PR DESCRIPTION
**Pull request by pvanmulbregt**
Attempt at adding universal dispatcher for the 1d root-solvers, analogous to `optimize.root()` for the vector functions. Adding testing by name using `root_scalar()`. Addresses gh-8884.

Added `x1=None`, `rtol=0` as optional arguments to `newton()`.
If `fprime` is not specified, and `x1` is, use that instead of `1.0001*x0`.
Allow specifying a `rtol` for the convergence criterion.
The signature is now: 
```
newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
        fprime2=None, x1=None, rtol=0.0,
        full_output=False, disp=True)
```
@pv  The `full_output` and `disp` which were added to `newton()` two weeks ago (gh-8803) fit naturally at the end so I placed the two new keywords just before them. For complete backwards compatibility, however, the new ones should be at the end.  Are such changes permitted before the code is incorporated into an actual release? Or is it the case that once a PR has been merged, anything future must be compatible with it.

### CODE REVIEW

**Comment by rgommers** [comment]
>  Are such changes permitted before the code is incorporated into an actual release? Or is it the case that once a PR has been merged, anything future must be compatible with it.

As long as code is not in a released version, it can still be changed. So inserting the new keywords like you did seems right.

**Comment by pvanmulbregt** [comment]
There are multiple open (or recently closed) PRs dealing with scalar root-finding.[ O=Open, C=Closed, R=to-be-replaced] 

- gh-8357 [O] Vectorize `newton`.
- gh-8431  [O] Cython optimize.
- gh-8761  [R] Add Sidi's root-finding algorithm.
- gh-8803  [C] additional optional `full_output` arg to `newton()`.
- gh-8835  [O] Replace Newton's method with Homeier's modification.
- gh-8876  [O] Add TOMS748 algorithm as root finder.  Enhanced test suite/framework.
- gh-8882  [O] Reduced (safer) use of Halley's algorithm.
- gh-8889  [O] Rework optimize top-level doc.
- gh-8890  [O] universal dispatcher for 1d root finders, optional `rtol`, `x1` args to `newton()`.
- (maybe some others I've missed ....  Is there a way to ask GitHub for all PRs that modify a specific file?)

Some thoughts when considering these PRs:
gh-8835 is not backwards compatible. Perhaps more naturally considered as a separate algorithm.
gh-8876 could naturally follow gh-8890. (actually the additional ~150 test cases and framework would be useful independent of adding TOMS748)
gh-8761's replacement could also naturally follow after gh-8890.
gh-8882 collides with most, but easily accommodates the others.
gh-8889 probably collides with gh-8890, but either could accommodate the other.
gh-8357 collides with gh-8890, but probably either could accommodate the other.
gh-8431 (Cythonizing) probably collides with everything.

**Comment by mikofski** [comment]
IMHO it would be easier to accommodate this PR and other new API changes (such as #8803) into #8357 if it were merged first. I greatly appreciate SciPy reviewers and maintainers time, but it has been in review since mid April.  I hope it can be merged soon. I have tried very hard to respond and adjust to many reviewers comments, and I believe it is a useful addition to SciPy. Perhaps @pvanmulbregt , @pv, or @rgommers might be interested in reviewing it? I understand you are very busy, but if you had any suggestions or concerns that I could address in order to merge it, I would be very grateful. Thank you!

**Comment by mikofski** [comment]
Also I plan to restart work on #8431 as soon as #8357 is merged. I had discussed this with @person142 a bit, and I also [posted on the scipy-dev mail list](https://mail.python.org/pipermail/scipy-dev/2018-April/022762.html). The idea is to create an API for users who desired tight looping as an alternate to #8357. The Cython optimize API would expose Brent, bisection, and Ridder in addition to Newton, but the user would have to write their own custom Cython code, so IMO they would be considered "outside" of this universal dispatcher for 1d root finders. For an customized Cython optimize example that uses `newton` and `bisect` see [`zeros_examples.pyx`](../../../mikofski/scipy/blob/cython_optimize/scipy/optimize/cython_optimize/tests/zeros_examples.pyx). 

PS - Sorry if I am spamming this PR. I can delete these comments if they are inappropriate here.

**Comment by rgommers** [comment]
@mikofski no worries, you're not spamming. We're just more short on reviewer manpower than usual - I know it's frustrating.

**Comment by pvanmulbregt** [comment]
@mikofski Actually, this is why I collected the open PRs and made few comments, hoping to have this discussion in order to understand possible interactions/implications before any of these are merged...

**Comment by pvanmulbregt** [comment]
One question that comes to mind when thinking about the 1-d root-finders is whether there is any underlying philosophy to the collection. 

The bracket-free derivative-using methods `newton/secant/halley` are not necessarily "root-finders" per se, but perhaps better considered as "root-polishers".   [It was striking to see with the test cases in  gh-8876 how many more function calls were needed by `newton` than say `brentq`.]

Currently `newton` is pretty minimal, failing under many circumstances under which it could recover. 
It could be made safer, by allowing a bracketing interval to be passed in by the user.   Or  more efficient by allowing the function to return `f` and its derivatives in a single call.  That starts to make the implementation more robust but also more complicated.  
Another viewpoint is that `newton` shouldn't really be recommended to find roots, but only to polish them, in situations where it will always work, and hence it should always have a bare-metal, user-beware implementation, forever.

**Comment by mikofski** [comment]
IMO the Newton/Halley/secant methods are supposed to be minimalist, and [the documentation explicitly states that they are not safe, that the bracketed methods are safer, and that `brentq` is preferred](https://github.com/scipy/scipy/blob/master/scipy/optimize/zeros.py). This is rendered in [the docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.newton.html#scipy.optimize.newton), and the preference for `brentq` is stated again in [the tutorial](https://docs.scipy.org/doc/scipy/reference/tutorial/optimize.html#scalar-functions).

I agree the that we _might_ be able to improve Newton slightly. I have suggested backtracking and line searching using previous guesses when convergence is poor as I did for my [MATLAB Newton-Raphson solver](https://www.mathworks.com/matlabcentral/fileexchange/43097-newton-raphson-solver) although maybe it's disallowed since I paraphrased from NR, and I'm still ambivalent.

So I'm general I'm in favor of your last statement, that `newton` be considered "bare-metal, user-beware implementation, forever."

File: scipy/optimize/_root_scalar.py
Comment (rgommers) [inline scipy/optimize/_root_scalar.py]: Would be useful to clarify here that the size of the memoized set is 1; i.e. this only does something when calling the function with the same `x` 2x in a row. And that it's assumed that `args` does not change between function calls.

File: scipy/optimize/_root_scalar.py
Comment (rgommers) [inline scipy/optimize/_root_scalar.py]: `bracket : sequence of 2 floats, optional`

File: scipy/optimize/_root_scalar.py
Comment (rgommers) [inline scipy/optimize/_root_scalar.py]: boolean should be lower-case (here and below)

File: scipy/optimize/_root_scalar.py
Comment (rgommers) [inline scipy/optimize/_root_scalar.py]: this whole section is a copy-paste error

File: scipy/optimize/_root_scalar.py
Comment (rgommers) [inline scipy/optimize/_root_scalar.py]: section should be left out if empty

File: scipy/optimize/_root_scalar.py
Comment (rgommers) [inline scipy/optimize/_root_scalar.py]: no blank line needed here

File: scipy/optimize/_root_scalar.py
Comment (rgommers) [inline scipy/optimize/_root_scalar.py]: need a blank line here

File: scipy/optimize/_root_scalar.py
Comment (rgommers) [inline scipy/optimize/_root_scalar.py]: a couple of PEP8 violations slipped in, best to fix those: `n_calls` here, and `is_memoized` further down

**Comment by rgommers** [comment]
Needs a rebase. Overall this looks quite good to me. Made lots of little comments, nothing major yet - still need to play with the code a bit.

**Comment by rgommers** [comment]
The tests won't run like this, still needs a rebase.

**Review by rgommers — CHANGES_REQUESTED** [review]
Some more comments on minor things. Pointer on how to fix the html docs is the most important of these.

**Comment by rgommers** [comment]
A test for `method=None` seems to be missing, trying that turns up an issue:
```
optimize.root_scalar(lambda x: x**2 - 2)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-358881f9acc8> in <module>()
----> 1 optimize.root_scalar(lambda x: x**2 - 2)

~/Code/scipy/scipy/optimize/_root_scalar.py in root_scalar(f, args, method, bracket, fprime, fprime2, x0, x1, xtol, rtol, maxiter, options)
    161         args = (args,)
    162 
--> 163     meth = method.lower()
    164     if options is None:
    165         options = {}

AttributeError: 'NoneType' object has no attribute 'lower'
```

**Comment by rgommers** [comment]
Second issue, when specifying a method but not `x0`:
```
In [9]: optimize.root_scalar(lambda x: x**2 - 2, method='newton')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-a285ea476781> in <module>()
----> 1 optimize.root_scalar(lambda x: x**2 - 2, method='newton')

~/Code/scipy/scipy/optimize/_root_scalar.py in root_scalar(f, args, method, bracket, fprime, fprime2, x0, x1, xtol, rtol, maxiter, options)
    218             kwargs['tol'] = kwargs.pop('xtol')
    219         r, sol = methodc(f, x0, args=args, fprime=fprime, fprime2=fprime2,
--> 220                          x1=x1, **kwargs)
    221     else:
    222         raise ValueError('Unknown solver %s' % method)

~/Code/scipy/scipy/optimize/zeros.py in newton(func, x0, fprime, args, tol, maxiter, fprime2, x1, rtol, full_output, disp)
    192     # Multiply by 1.0 to convert to floating point.  We don't use float(x0)
    193     # so it still works if x0 is complex.
--> 194     p0 = 1.0 * x0
    195     funcalls = 0
    196     if fprime is not None:

TypeError: unsupported operand type(s) for *: 'float' and 'NoneType'
```

**Comment by rgommers** [comment]
Third issue:
```
In [11]: optimize.root_scalar(lambda x: x**2 - 2, method='brentq', x0=1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-8b5255d59a85> in <module>()
----> 1 optimize.root_scalar(lambda x: x**2 - 2, method='brentq', x0=1)

~/Code/scipy/scipy/optimize/_root_scalar.py in root_scalar(f, args, method, bracket, fprime, fprime2, x0, x1, xtol, rtol, maxiter, options)
    212 
    213     if meth in ['bisect', 'ridder', 'brentq', 'brenth']:
--> 214         a, b = bracket[:2]
    215         r, sol = methodc(f, a, b, args=args, **kwargs)
    216     elif meth == 'newton':

TypeError: 'NoneType' object is not subscriptable
```

Generally seems to need some checks for "expected user errors".

**Comment by rgommers** [comment]
When I do things as intended, all seems fine. Nice interface.

**Comment by pvanmulbregt** [comment]
On @rgommers  3 issues:
1. A `None` method is _supposed_ to be fine.  In that case it selects which to use based on what data is provided: bracket => brentq, no bracket but f' => newton.  Otherwise raise ValueError.  The `meth = method.lower()` was incorrectly placed in all code paths.
2. The line `p0 = 1 * x0` has actually been present in `optimize.py:newton()` for at least 8 years.  I'll add a check in  `root_scalar` for `None`.
3. I'll add a check on bracket being a `list, tuple, np.ndarray`.

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
Changes for user errors look good, tried some more and all is well from what I can tell. 

Clicked approve on my review from yesterday too early - the doc build and other minor errors still need addressing.

**Comment by mikofski** [comment]
I will try to take at this over the weekend if that's helpful.

File: scipy/optimize/_root_scalar.py
Comment (mikofski) [inline scipy/optimize/_root_scalar.py]: "use" - present tense?

ditto for `fprime` and `fprime2` below

File: scipy/optimize/_root_scalar.py
Comment (mikofski) [inline scipy/optimize/_root_scalar.py]: perhaps `fun`? Since this #8890 is supposed to address #8884 "inconsistent names of function arguments" then shouldn't `root_scalar` use the same keyword arguments as `minimize`, `minimize_scalar`, and `root` which all use `fun`? Also I think in general (and there are always exceptions) pep-8 discourages the use of single letter variables names

okay, I see, all of the scalar roots except newton are currently using `f` so maybe that is the prevailing convention. I'm fine either way.

File: scipy/optimize/__init__.py
Comment (mikofski) [reply scipy/optimize/__init__.py]: I was able to get the docs to show up by adding these files to `scipy/doc/source`

* `optimize.root_scalar-brentq.rst`

```rst
.. _optimize.root_scalar-brentq:

root_scalar(method='brentq')
----------------------------------------------

.. scipy-optimize:function:: scipy.optimize.root_scalar
   :impl: scipy.optimize.zeros.brentq
   :method: brentq
```

etc.

File: scipy/optimize/optimize.py
Comment (mikofski) [reply scipy/optimize/optimize.py]: Hmm, that didn't work. I add all 5 files to `scipy/doc/source`, and `show_options` still doesn't include them
[optimize.root_scalar-method.rst.zip](https://github.com/scipy/scipy/files/2151521/optimize.root_scalar-method.rst.zip)

File: scipy/optimize/_root_scalar.py
Comment (mikofski) [inline scipy/optimize/_root_scalar.py]: perhaps drop `fprime=True` since if `fprime2` is true, then it doesn't matter what `fprime` is? Or perhaps add a note that reiterates the docstring for `fprime2`?

File: scipy/optimize/_root_scalar.py
Comment (mikofski) [inline scipy/optimize/_root_scalar.py]: Can you please add a "Notes" section that explains that if `bracket` is given, but not `method`, then the default is `brentq`, and if no `bracket` but `x0` then it's `newton`? The "Notes" for [`minimize_scalar`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize_scalar.html) are more verbose than this.

Also maybe in a future PR it might be nice to just make the default `brentq` and have `root_scalar` [get a bracket like `Brent` does](../blob/14142ff70d84a6ce74044a27919c850e893648f7/scipy/optimize/optimize.py#L1863-L1869) by calling [`bracket`](blob/v1.1.0/scipy/optimize/optimize.py#L2256)?

File: scipy/optimize/tests/test_zeros.py
Comment (mikofski) [inline scipy/optimize/tests/test_zeros.py]: perhaps use [`TOL` from line 18](../../blob/8b1addf06136a47280eaabb52a7e5e8113f6c6b7/scipy/optimize/tests/test_zeros.py#L18)?

File: scipy/optimize/zeros.py
Comment (mikofski) [inline scipy/optimize/zeros.py]: would you consider using `np.isclose` instead?

File: scipy/optimize/zeros.py
Comment (mikofski) [inline scipy/optimize/zeros.py]: I apologize for my ignorance, but I don't understand the difference between these two cases, they always return the same values for me:

```python
>>> f = lambda x: x**3 - 2.
>>> x1, x2 = 1.3, 1.5
>>> f1, f2 = f(x1), f(x2)
>>> x1, f1, x2, f2
(1.3, 0.19700000000000006, 1.5, 1.375)

# -> abs(f1) < abs(f2)

>>> (f1*x2 - f2*x1) / (f1 - f2)
1.266553480475382

>>> (f2*x1 - f1*x2) / (f2-f1)
1.266553480475382

# both conditions return the same "x"

>>> x1, x2 = -1.3, 1.5
>>> f1, f2 = f(x1), f(x2)
>>> x1, f1, x2, f2
>>> (-1.3, -4.197, 1.5, 1.375)

# -> abs(f1) > abs(f2)

>>> (f1*x2 - f2*x1) / (f1 - f2)
0.8090452261306534

>>> (f2*x1 - f1*x2) / (f2-f1)
0.8090452261306534

# both conditions return the same "x"

>>> x1, x2 = 1.3, -1.5
>>> f1, f2 = f(x1), f(x2)
>>> x1, f1, x2, f2
(1.3, 0.19700000000000006, -1.5, -5.375)

# -> abs(f1) < abs(f2)

>>> (f1*x2 - f2*x1) / (f1 - f2)
1.2010050251256281

>>> (f2*x1 - f1*x2) / (f2-f1)
1.2010050251256281

# both conditions return the same "x"

>>> x1, x2 = -1.3, -1.5
>>> f1, f2 = f(x1), f(x2)
>>> x1, f1, x2, f2
(-1.3, -4.197, -1.5, -5.375)

# -> abs(f2) > abs(f1)

>>> (f1*x2 - f2*x1) / (f1 - f2)
-0.5874363327674018

>>> (f2*x1 - f1*x2) / (f2-f1)
-0.5874363327674018

# both conditions return the same "x"
```

is this missing `sign()`? I remember you had a comment about the slope going the wrong way, but now I can't find it anywhere.

**Review by mikofski — APPROVED** [review]
Thanks @pvanmulbregt for adding this contribution and improving the scalar roots!

File: scipy/optimize/optimize.py
Comment (mikofski) [reply scipy/optimize/optimize.py]: You also need to add this to `scipy/optimize/optimize.py`:

```patch
diff --git a/scipy/optimize/optimize.py b/scipy/optimize/optimize.py
index bc4925e08..374453de5 100644
--- a/scipy/optimize/optimize.py
+++ b/scipy/optimize/optimize.py
@@ -2936,6 +2936,15 @@ def show_options(solver=None, method=None, disp=True):
     - :ref:`krylov            <optimize.root-krylov>`
     - :ref:`df-sane           <optimize.root-dfsane>`
 
+    `scipy.optimize.root_scalar`
+
+    - :ref:`brentq       <optimize.root_scalar-brentq>`
+    - :ref:`brenth       <optimize.root_scalar-brenth>`
+    - :ref:`ridder       <optimize.root_scalar-ridder>`
+    - :ref:`bisect       <optimize.root_scalar-bisect>`
+    - :ref:`newton       <optimize.root_scalar-newton>`
+
+
     `scipy.optimize.minimize_scalar`
 
     - :ref:`brent       <optimize.minimize_scalar-brent>`
@@ -3003,6 +3012,9 @@ def show_options(solver=None, method=None, disp=True):
         text.extend(["\n\n\n====\n", "root\n",
                      "====\n"])
         text.append(show_options('root', disp=False))
+        text.extend(["\n\n\n====\n", "root_scalar\n",
+                     "====\n"])
+        text.append(show_options('root_scalar', disp=False))
         text.extend(['\n\n\n=======\n', 'linprog\n',
                      '=======\n'])
         text.append(show_options('linprog', disp=False))
```

now it works!

File: scipy/optimize/zeros.py
Comment (pvanmulbregt) [reply scipy/optimize/zeros.py]: Yes.  Both expressions attempt to compute the mathematical `(q1 p0 - q0 p1)/(q1 - q0)`. If `|q1/q0| > 1e16` then the contribution of `q0` can be lost.  So start from the endpoint with the smallest value of the function and move a slight distance away `p0 + epsilon`, rather than from the other end where one should really move `(p0-p1)+epsilon`, but `epsilon` is swamped by `p0-p1`. [Yes, some test cases really do have variation this large.]

File: scipy/optimize/__init__.py
Comment (pvanmulbregt) [reply scipy/optimize/__init__.py]: Agree.

File: scipy/optimize/_root_scalar.py
Comment (pvanmulbregt) [reply scipy/optimize/_root_scalar.py]: Agree.

File: scipy/optimize/_root_scalar.py
Comment (pvanmulbregt) [reply scipy/optimize/_root_scalar.py]: I'm not sure I want to hard-code the decision for all-time. It's an interesting issue.
I'm thinking ahead to two current /future PRs.
gh-8876 adds TOMS 748 which may be judged a better default for bracketed toot-finding.
Sidi's method would expand to cover complex-valued functions in the complex domain when given two points.
So other default would be "use the best method available for the situation presented".
Thoughts?

File: scipy/optimize/__init__.py
Comment (pvanmulbregt) [reply scipy/optimize/__init__.py]: I based `root_scalar` on `root`, but there is a difference.  `root` accepts a couple of method-specific options as known keywords and requires the rest to be presented as part of an `options` dict.  `root_scalar` is allowing these options to be presented as keywords.  E.g.
```
def root(fun, x0, args=(), method='hybr', jac=None, tol=None, callback=None, options=None):
vs
def root_scalar(f, args=(), method=None, bracket=None,
                fprime=None, fprime2=None, x0=None, x1=None,
                xtol=None, rtol=None, maxiter=None, options=None):
```

Even `maxiter` is method-specific in `root`, requiring the user to specify 
```
root(fun, ..., options={'maxiter': 100, 'disp: True})
```
whereas the 1-d usage has been 
```
newton(f, maxiter=100, disp=True)
```
To keep with then 1-d usage, especially as selecting a method based on inputs, I kept these options as top-level keyword arguments.  It also allows mapping between `tol` and `xtol` and `atol` transparently to the user, which are inconsistent in the 1-d methods.
Currently there are no 1-d solvers that take any additional options. (Sidi's method would have one option.)  So I'm looking at whether it is worthwhile to exactly match `root` in this regard, especially with `optimize.show_options()`, or whether to deviate slightly.

File: scipy/optimize/_root_scalar.py
Comment (rgommers) [reply scipy/optimize/_root_scalar.py]: > So other default would be "use the best method available for the situation presented".

this makes sense to me

Comment (mikofski) [reply scipy/optimize/_root_scalar.py]: Personally I would steer users away from Newton unless `fprime` is given. If `x0` is given but not `fprime`, then I would suggest growing a bracket around `x0` and calling Brent. Secant is not reliable IMO. For example try to solve the single diode model with a bad guess and you get GIGO.

```python
IL, I0, RS, RSH, GAMMA, VT = 7., 1.e-6, 0.01, 20., 1.23, 0.026

def sdm(v, i, il, i0, rs, rsh, gamma,vt):
    vd=v+I*rs
    f = il-i0*(np.exp(vd/gamma/vt)-1)-vd/rsh-i
    return f

VOC=GAMMA*np.log(IL/I0+1.)
I=3.  # current [A]
ARGS=(I, IL, I0, RS, RSH, GAMMA, VT)
V0 = 0.3. # voltage [V] bad guess

root_scalar(sdm, bracket=(0,VOC))

# works!

root_scalar(sdm, x0=VOC)

# also works!

root_scalar(sdm, x0=V0). # bad guess

# fails!
```

**Comment by pvanmulbregt** [comment]
Split `root_scalar(f, method='newton')` into 3 constituent cases. `method='newton'` (requires 1 derivative), `method='halley'` (requires two derivatives `fprime`and `fprime2`), `method='secant'` (no derivatives). This way the user is being explicit as to which approach they want.  
[If `method` is None, then `root_scalar` will still select the most appropriate method.
If the user calls `optimize.newton` directly, there is no change in behavior. 
This modification is only about getting a clean interface for `root_scalar()`.]
Thoughts?

**Comment by rgommers** [comment]
> Split `root_scalar(f, method='newton')` into 3 constituent cases.

This sounds like a good idea to me. The method names don't have to be 1:1 corresponding with current functions; it should make sense stand-alone and it does here. These are 3 related but separate algorithms, so can be split in the UX here.

**Comment by mikofski** [comment]
cross referencing #5504 an earlier attempt abandoned by @person142

File: scipy/optimize/tests/test_zeros.py
Comment (mikofski) [reply scipy/optimize/tests/test_zeros.py]: Why not use `TOL` same value, sound it's already defined on L21? Sorry to comment so late, feel free to ignore. :)

Comment (rgommers) [reply scipy/optimize/tests/test_zeros.py]: thanks @mikofski, incorrectly resolved this one. it's a tiny thing so I'm going to spare the CI another re-run, apologies

**Pull request merged**
